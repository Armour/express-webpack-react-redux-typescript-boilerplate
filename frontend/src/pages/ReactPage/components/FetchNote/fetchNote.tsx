import { List } from 'immutable';
import React from 'react';
import { Translation } from 'react-i18next';
import { connect } from 'react-redux';
import { AnyAction, Dispatch } from 'redux';

import { ADD_NOTE_REQUESTED, EDIT_NOTE_REQUESTED, FETCH_ALL_NOTES_REQUESTED, FETCH_NOTE_REQUESTED, REMOVE_NOTE_REQUESTED } from 'services/notes/constants';
import { INote } from 'services/notes/types';
import { IGlobalStateRecord } from 'types/global';

const styles = require('./fetchNote.scss');

// Component

interface IFetchNoteStateProps {
  notes: List<INote>;
  loading: boolean;
  error: string;
}

interface IFetchNoteDispatchProps {
  fetchAllNotes(): void;
  fetchNote(id: number): void;
  addNote(content: string): void;
  editNote(id: number, content: string): void;
  removeNote(id: number): void;
}

interface IFetchNoteProps extends IFetchNoteStateProps, IFetchNoteDispatchProps { }

export class FetchNote extends React.Component<IFetchNoteProps> {
  public fetchAllNotes = (e: React.MouseEvent<HTMLElement>) => {
    e.preventDefault();
    this.props.fetchAllNotes();
  }

  public render() {
    const noteList = this.props.notes.map((note) =>
      (
        <li className='collection-item'>
          <div>Id: {note.id} &emsp; Content: {note.content}</div>
        </li>
      ),
    );
    const noteListCollection = <ul className={`collection ${styles['fetch-note-collection']}`}> {noteList} </ul>;
    const errorPanel = (
      <div className={`card-panel red ${styles['fetch-note-error-panel']}`}>
        <span className='white-text'>{this.props.error}</span>
      </div>
    );
    return (
      <Translation ns='reactPage'>
        {(t) => (
          <div className={`center-align z-depth-2 ${styles['fetch-note-layout']}`}>
            <span className={styles['fetch-note-title']}>{t('fetchNote.asyncCalls')}</span>
            <div className='fetchAllNotes'>
              <a className={`btn waves-effect ${styles['fetch-note-filter-btn']}`} onClick={this.fetchAllNotes} role='button'>
                {t('fetchNote.fetchAllNotes')}
              </a>
            </div>
            {noteList.count() > 0 && noteListCollection}
            {this.props.error !== '' && errorPanel}
          </div>
        )}
      </Translation>
    );
  }
}

// Container

const mapStateToProps = (state: IGlobalStateRecord): IFetchNoteStateProps => ({
    notes: state.get('notesState').notes,
    loading: state.get('notesState').loading,
    error: state.get('notesState').error,
});

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>): IFetchNoteDispatchProps => ({
  fetchAllNotes: () => {
    dispatch({ type: FETCH_ALL_NOTES_REQUESTED, payload: {} });
  },
  fetchNote: (id: number) => {
    dispatch({ type: FETCH_NOTE_REQUESTED, payload: { id } });
  },
  addNote: (content: string) => {
    dispatch({ type: ADD_NOTE_REQUESTED, payload: { content } });
  },
  editNote: (id: number, content: string) => {
    dispatch({ type: EDIT_NOTE_REQUESTED, payload: { id, content } });
  },
  removeNote: (id: number) => {
    dispatch({ type: REMOVE_NOTE_REQUESTED, payload: { id } });
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FetchNote);
