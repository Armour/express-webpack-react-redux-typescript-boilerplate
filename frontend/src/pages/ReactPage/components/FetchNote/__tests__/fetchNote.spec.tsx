import { List } from 'immutable';
import 'materialize-css';
import React from 'react';
import TestRenderer from 'react-test-renderer';

import { FetchNote } from '../fetchNote';

// Mock dispatch
const dispatchMock = () => { return; };

describe('FetchNote', () => {
  it('should renders correctly', () => {
    const renderer = TestRenderer.create(
      <FetchNote
        notes={List<any>()}
        loading={false}
        error={''}
        fetchAllNotes={dispatchMock}
        fetchNote={dispatchMock}
        addNote={dispatchMock}
        editNote={dispatchMock}
        removeNote={dispatchMock}
      />,
    );
    expect(renderer).toMatchSnapshot();
    renderer.unmount();
  });
});
