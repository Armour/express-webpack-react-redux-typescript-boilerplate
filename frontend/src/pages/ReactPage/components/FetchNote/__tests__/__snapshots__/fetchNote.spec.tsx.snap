// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchNote should renders correctly 1`] = `
<div
  className="center-align z-depth-2 fetch-note-layout"
>
  <span
    className="fetch-note-title"
  >
    fetchNote.asyncCalls
  </span>
  <div
    className="fetchAllNotes"
  >
    <a
      className="btn waves-effect fetch-note-filter-btn"
      onClick={[Function]}
      role="button"
    >
      fetchNote.fetchAllNotes
    </a>
  </div>
</div>
`;
